compiler/rustc_query_system/src/dep_graph/serialized.rs

```diff
@@ -152,6 +193,31 @@ impl SerializedDepGraph {
     pub fn node_count(&self) -> usize {
         self.nodes.len()
     }
+
+    /// This spawns a thread that prefetches the index.
+    fn prefetch(self: &Arc<Self>) {
+        if !self.index.is_empty() {
+            let this = self.clone();
+            thread::spawn(move || {
```

Comment:
Shouldn't we block this on a job server token being available for this extra computation work?

Reply:
I'm not sure if that's performance win as it's more efficient if this completes in a timely manner. `setup_index` is less efficient than doing all the dep kinds at once.

Reply:
> I'm not sure if that's performance win as it's more efficient if this completes in a timely manner.

I don't really follow. Jobserver token availability isn't about performance, strictly speaking, it's about making sure that we're not consuming more resources than the host has and/or the user is willing to give. We've definitely had complaints about -j1 (for example) not being respected before.

If we don't have a token available, that may mean that we should do the work in-band (i.e., not spawning the thread) even if that is slower. But, that's what's going to happen anyway on a system that's already CPU-saturated - just via kernel scheduling - which is the alternative here, right? So I don't really understand how the token would be a problem.

Reply:
I guess we could add a `try_acquire` method to the jobserver and only spawn the thread if we get a token. It looks like that would be racy on POSIX though. I'm not sure if macOS or Linux offers a way to do non-blocking reads.

Reply:
I've made the PR check for a free token now. It's a bit racy but probably works fine.